poly/promise/promise.go

```diff
@@ -0,0 +1,1355 @@
+
+	// A list of servers who will act as insurers of the Promise. The list
+	// contains the long-term public keys of the insurers
+	insurers []abstract.Point
+
+	// The list of shared secrets to be sent to the insurers. They are
+	// encrypted with Diffie-Hellman shared secrets between the insurer
+	// and the promiser.
+	secrets []abstract.Secret
+}
+
+/* Constructs a new Promise to guarentee a secret.
+ *
+ * Arguments
+ *    secretPair   = the keypair of the secret to be promised
+ *    longPair     = the long term keypair of the promiser
+ *    t            = minimum number of shares needed to reconstruct the secret.
+ *    r            = minimum PromiseSignatures needed to certify the Promise
+ *    insurers     = a list of the long-term public keys of the insurers.
+ *
+ * Returns
+ *   A newly constructed Promise
+ */
+func (p *Promise) ConstructPromise(secretPair *config.KeyPair,
+	longPair *config.KeyPair, t, r int, insurers []abstract.Point) *Promise {
+	p.id = secretPair.Public.String()
+	p.t = t
+	p.r = r
+	p.n = len(insurers)
+	p.shareSuite = secretPair.Suite
+	p.keySuite = longPair.Suite
+	p.pubKey = longPair.Public
+	p.insurers = insurers
+	p.secrets = make([]abstract.Secret, p.n, p.n)
+
+	// Verify that t <= r <= n
+	if p.n < p.t {
+		panic("Not enough insurers for the secret")
+	}
+	if p.r < p.t {
+		p.r = p.t
+	}
+	if p.r > p.n {
+		p.r = p.n
+	}
+
+	// Create the public polynomial and private shares. The number of shares
+	// should be equal to the number of insurers.
+	pripoly := new(poly.PriPoly).Pick(p.shareSuite, p.t,
+		secretPair.Secret, random.Stream)
+	prishares := new(poly.PriShares).Split(pripoly, p.n)
+	p.pubPoly = poly.PubPoly{}
+	p.pubPoly.Commit(pripoly, nil)
+
+	// Populate the secrets array with the shares encrypted by a Diffie-
+	// Hellman shared secret between the promiser and appropriate insurer
+	for i := 0; i < p.n; i++ {
+		diffieBase := p.keySuite.Point().Mul(insurers[i], longPair.Secret)
+		p.secrets[i] = p.diffieHellmanEncrypt(prishares.Share(i), diffieBase)
+	}
+	return p
+}
+
+/* Initializes a Promise for unmarshalling
+ *
+ * Arguments
+ *    shareSuite = the suite used for shared secrets within the Promise
+ *    keySuite   = the suite used for long term keys within the Promise
+ *
+ * Returns
+ *   An initialized Promise ready to be unmarshalled
+ */
+func (p *Promise) UnmarshalInit(shareSuite, keySuite abstract.Suite) *Promise {
+	p.shareSuite = shareSuite
+	p.keySuite = keySuite
+	return p
+}
+
+/* Verifies that the Promise was constructed correctly.
+ *
+ * Arguments
+ *    promiserKey = the long term key the caller believes the Promise to be from
+ *
+ * Return
+ *   an error if the promise is malformed, nil otherwise.
+ *
+ * TODO Consider more ways to verify (such as making sure there are no duplicate
+ *      keys in p.insurers or that the promiser's long term public key is not in
+ *      p.insurers).
+ */
+func (p *Promise) VerifyPromise(promiserKey abstract.Point) error {
```

Comment:
The UnmarshalBinary function should call this automatically and fail if a received promise is invalid.  If that's happening, then maybe this can just be an internal helper and no longer needs to be a public method?